src/context/AppContext.tsx

```diff
@@ -0,0 +1,106 @@
+import { createContext, useCallback, useEffect, useState } from 'react';
+import { Product } from '../types/ProductCard';
+import { CartProducts } from '../types/CartProducts';
+
+export type CartContextType = {
+  cartProducts: CartProducts;
+  addToCart: (product: Product) => void;
+  incrementProductInCart: (productId: Product['id']) => void;
+  decrementProductInCart: (productId: Product['id']) => void;
+  removeFromCart: (productId: Product['id']) => void;
+  clearCart: () => void;
+};
+
+export const CartContext = createContext<CartContextType>({
+  cartProducts: {},
+  addToCart: () => {},
+  incrementProductInCart: () => {},
+  decrementProductInCart: () => {},
+  removeFromCart: () => {},
+  clearCart: () => {},
+});
+
+export const AppContext: React.FC<{ children: React.ReactNode }> = ({
+  children,
+}) => {
+  const cartFromLocalStorage = localStorage.getItem('cart');
```

Comment:
As an example of a custom hook you may use here: 

```
export const useLocalStorage = (key, initialValue) => {
  const [storedValue, setStoredValue] = useState(() => {
    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.log(error);
      return initialValue;
    }
  });
```

As you can see:
1. It's safer since you have `try/catch` to work with errors without app crash,
2. It's easier to provide an `initialValue`,
3. It's just easier to use all around your app :)